### PR TITLE
Add check for keywords in PR titles to block merge

### PR DESCRIPTION
**Pull request by corbin-coleman**
Add another acceptable configuration keyword, `merge_block_keywords` that can be added to the `.unir.yml` file. If the PR title contains any of the words in the `merge_block_keywords` section then the PR will not be able to automatically be able to merge.

A good example of usage would be with `WIP` PRs. They shouldn't be merged until they are no longer a `WIP`, but they may get an appropriate number of `LGTM`s before that happens.

The keywords are defaulted to `WIP:` if no keywords are defined in the `.unir.yml` file.

Signed-off-by: corbin-coleman <corbin.coleman@docker.com>

### CODE REVIEW

**Comment by jose-bigio** [comment]
Here's the build error

```
internal/webhook.go:252:5: undefined: reviewEvent
internal/webhook.go:269:4: too many arguments to return
	have (bool)
	want ()
internal/webhook.go:272:2: too many arguments to return
	have (bool)
	want ()
internal/webhook.go:326:28: checkMergeBlockKeywords(config.MergeBlockKeywords, prTitle) used as value
make: *** [build/unir] Error 2
```

File: internal/webhook.go [hunk 1]
```diff
@@ -249,17 +249,30 @@ func handleStatus(integrationID int, keyfile string, statusEvent github.StatusEv
 			*statusEvent.Repo.Name,
 			*pullRequest.Head.SHA,
 			*pullRequest.Number,
+			*reviewEvent.PullRequest.Title,
```
Comment (jose-bigio) [inline internal/webhook.go]: I'm guess this should be 
*pullRequest.Title

File: internal/webhook.go [hunk 2]
```diff
@@ -249,17 +249,30 @@ func handleStatus(integrationID int, keyfile string, statusEvent github.StatusEv
...
 
 func handlePullRequestReview(integrationID int, keyfile string, reviewEvent github.PullRequestReviewEvent) {
 	client := createGithubClient(integrationID, int(*reviewEvent.Installation.ID), keyfile)
 	log.Debugf("[%s] STARTED handling pull request review", *reviewEvent.Review.HTMLURL)
-	mergePullRequest(client, *reviewEvent.Repo.Owner.Login, *reviewEvent.Repo.Name, *reviewEvent.PullRequest.Head.SHA, *reviewEvent.PullRequest.Number)
+	mergePullRequest(client, *reviewEvent.Repo.Owner.Login, *reviewEvent.Repo.Name, *reviewEvent.PullRequest.Head.SHA, *reviewEvent.PullRequest.Number, *reviewEvent.PullRequest.Title)
 }
 
-func mergePullRequest(client *github.Client, owner, repo, sha string, prNumber int) {
+func checkMergeBlockKeywords(mergeBlockKeywords []string, prTitle string) {
```
Comment (jose-bigio) [inline internal/webhook.go]: The function signature doesn't indicate that a bool is being returned (which is where the other errors are coming from).

File: internal/config.go [hunk 3]
```diff
@@ -5,10 +5,11 @@ import (
...
 type UnirConfig struct {
-	Whitelist       []string `yaml:"whitelist"`
-	ApprovalsNeeded int      `yaml:"approvals_needed"`
-	ConsensusNeeded bool     `yaml:"consensus_needed"`
-	MergeMethod     string   `yaml:"merge_method"`
+	Whitelist          []string `yaml:"whitelist"`
+	ApprovalsNeeded    int      `yaml:"approvals_needed"`
+	ConsensusNeeded    bool     `yaml:"consensus_needed"`
+	MergeMethod        string   `yaml:"merge_method"`
+	MergeBlockKeywords []string `yaml:"merge_block_keywords"`
```
Comment (seemethere) [inline internal/config.go:12]: ~Can we also add a test for this?~

Just to make sure we can actually read these in.

**Review by seemethere — CHANGES_REQUESTED** [review]

**Comment by jose-bigio** [comment]
To deploy we will need to create a local webhook handler. 
This handler will be something similar to: 
https://github.com/seemethere/unir/blob/master/internal/webhook.go#L23-#L32

In particular we will want to implement something like this: 
https://github.com/docker/batterup/blob/master/internal/webhook.go#L17-#L26 

Note that for the above code block an api token is utilized as opposed to an integration id. 
This will enable us to test `unir` locally without having to worry about setting up an app.  In order to toggle between these routers will need to add a flag in the cli to specify whether  we want the app based router or the webhook based router.

**Comment by jose-bigio** [comment]
Partial illustration of this working: 

https://github.com/jose-bigio/Webhooks/pull/28

File: internal/webhook.go [hunk 4]
```diff
@@ -184,20 +200,30 @@ func GrabConfig(ctx context.Context, client *github.Client, repo, owner string,
...
-		log.Fatal(err)
+func createGithubClient(integrationID, installationID int, keyfile, apiToken string, testClient bool) *github.Client {
+	if testClient {
+		log.Debug("Creating a client based on test")
+		ctx := context.Background()
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: apiToken},
+		)
+		return github.NewClient(oauth2.NewClient(ctx, ts))
+	} else {
```
Comment (golangci) [inline internal/webhook.go]: if block ends with a return statement, so drop this else and outdent its block

File: cmd/unir/main.go [hunk 5]
```diff
@@ -18,22 +19,40 @@ var (
...
+	id := (os.Getenv(INTEGRATION_ID))
+
 	debug := flag.Bool("debug", false, "Toggle debug mode")
+	test := flag.Bool("test", false, "Toggle for testing")
 	port := flag.String("port", "8080", "Port to bind unir to")
 	flag.Parse()
+
+	// Use either a test handler which allows you to use a webhook
+	// or the regular handler which is integrated with an app
+	router := mux.NewRouter()
```
Comment (golangci) [inline cmd/unir/main.go]: ineffectual assignment to `router`

File: cmd/unir/main.go [hunk 6]
```diff
@@ -18,22 +19,40 @@ var (
...
-	}
-	router := internal.NewWebhookHandler(
-		[]byte(os.Getenv(WEBHOOK_SECRET)),
-		id,
-		os.Getenv(KEYFILE),
-	)
+	id := (os.Getenv(INTEGRATION_ID))
+
 	debug := flag.Bool("debug", false, "Toggle debug mode")
+	test := flag.Bool("test", false, "Toggle for testing")
```
Comment (seemethere) [inline cmd/unir/main.go]: We should just rename this to `oauth` mode

File: internal/webhook.go [hunk 7]
```diff
@@ -310,6 +360,11 @@ func mergePullRequest(client *github.Client, owner, repo, sha string, prNumber i
 		return
 	}
 
+	if checkMergeBlockKeywords(config.MergeBlockKeywords, prTitle) {
+		doStatus("checking keywords that block merging", "failure", "Can't merge currently, title contains keywords that block merges.")
```
Comment (seemethere) [inline internal/webhook.go]: I think the commit context should read:

> Automatic merging blocked, title contains keywords that prevent unir from automatically merging

File: internal/webhook.go [hunk 8]
```diff
@@ -12,19 +12,35 @@ import (
...
+	router.Handle("/", http.HandlerFunc(handler.handleGithubWebhook)).Methods("POST")
+	return router
 }
 
 func NewWebhookHandler(secret []byte, integrationID int, keyfile string) *mux.Router {
 	handler := GithubWebhookHandler{
 		Secret:        secret,
 		integrationID: integrationID,
 		keyfile:       keyfile,
+		testClient:    false,
```
Comment (seemethere) [inline internal/webhook.go]: Should be renamed to `needsOAuthClient`, since that's literally what it does.

File: cmd/unir/main.go (hunk 5 above)
Comment (seemethere) [reply cmd/unir/main.go]: I agree with @golangci here, this doesn't need to be assigned here

**Review by seemethere — CHANGES_REQUESTED** [review]

**Comment by jose-bigio** [comment]
It's green!

**Review by seemethere — APPROVED** [review]
LGTM

**Pull request merged**